image/image.go

```diff
@@ -0,0 +1,386 @@
+	Snaps            []string
+	Rootdir          string
+	Channel          string
+	ModelAssertionFn string
+	GadgetUnpackDir  string
+}
+
+func Prepare(opts *Options) error {
+	if err := downloadUnpackGadget(opts); err != nil {
+		return err
+	}
+
+	return bootstrapToRootdir(opts)
+}
+
+func decodeModelAssertion(fn string) (*asserts.Model, error) {
+	rawAssert, err := ioutil.ReadFile(fn)
+	if err != nil {
+		return nil, fmt.Errorf("cannot read model assertion: %s", err)
+	}
+
+	ass, err := asserts.Decode(rawAssert)
+	if err != nil {
+		return nil, fmt.Errorf("cannot decode model assertion %q: %s", fn, err)
+	}
+	return ass.(*asserts.Model), nil
+}
+
+func downloadUnpackGadget(opts *Options) error {
+	model, err := decodeModelAssertion(opts.ModelAssertionFn)
+	if err != nil {
+		return err
+	}
+	if err := os.MkdirAll(opts.GadgetUnpackDir, 0755); err != nil {
+		return fmt.Errorf("cannot create gadget unpack dir %q: %s", opts.GadgetUnpackDir, err)
+	}
+
+	dlOpts := &downloadOptions{
+		TargetDir:    opts.GadgetUnpackDir,
+		Channel:      opts.Channel,
+		StoreID:      model.Store(),
+		Architecture: model.Architecture(),
+	}
+	snapFn, _, err := acquireSnap(model.Gadget(), dlOpts)
+	if err != nil {
+		return err
+	}
+	// FIXME: jumping through layers here, we need to make
+	//        unpack part of the container interface (again)
+	snap := squashfs.New(snapFn)
+	return snap.Unpack("*", opts.GadgetUnpackDir)
+}
+
+func acquireSnap(snapName string, dlOpts *downloadOptions) (string, *snap.Info, error) {
+	if osutil.FileExists(snapName) {
+		return copyLocalSnapFile(snapName, dlOpts.TargetDir)
+	}
+
+	return downloadSnapWithSideInfo(snapName, dlOpts)
+}
+
+func bootstrapToRootdir(opts *Options) error {
+	if opts.Rootdir != "" {
+		dirs.SetRootDir(opts.Rootdir)
+		defer dirs.SetRootDir("/")
+	}
+
+	// sanity check target
+	if osutil.FileExists(dirs.SnapStateFile) {
+		return fmt.Errorf("cannot bootstrap over existing system")
+	}
+
+	model, err := decodeModelAssertion(opts.ModelAssertionFn)
+	if err != nil {
+		return err
+	}
+
+	// put snaps in place
+	if err := os.MkdirAll(dirs.SnapBlobDir, 0755); err != nil {
+		return err
+	}
+
+	snapSeedDir := filepath.Join(dirs.SnapSeedDir, "snaps")
+	dlOpts := &downloadOptions{
+		TargetDir:    snapSeedDir,
+		Channel:      opts.Channel,
+		StoreID:      model.Store(),
+		Architecture: model.Architecture(),
+	}
+
+	// FIXME: support sideloading snaps by copying the boostrap.snaps
+	//        first and keeping track of the already downloaded names
+	snaps := []string{}
+	snaps = append(snaps, opts.Snaps...)
+	snaps = append(snaps, model.Gadget())
+	snaps = append(snaps, model.Core())
+	snaps = append(snaps, model.Kernel())
+	snaps = append(snaps, model.RequiredSnaps()...)
+
+	for _, d := range []string{dirs.SnapBlobDir, snapSeedDir} {
+		if err := os.MkdirAll(d, 0755); err != nil {
+			return err
+		}
+	}
+
+	var seedYaml snap.Seed
+	for _, snapName := range snaps {
+		fmt.Printf("Fetching %s\n", snapName)
```

Comment:
Thanks good point. I added "image.Stdout" (similar to what we do in cmd/snap/main.go) to fix this.